flink-end-to-end-tests/test-scripts/test_resume_externalized_checkpoints.sh

```diff
@@ -24,114 +24,125 @@ fi
-    --test.semantics exactly-once \
-    --environment.parallelism $dop \
-    --environment.externalize_checkpoint true \
-    --environment.externalize_checkpoint.cleanup retain \
-    --state_backend $STATE_BACKEND_TYPE \
-    --state_backend.checkpoint_directory $CHECKPOINT_DIR_URI \
-    --state_backend.file.async $STATE_BACKEND_FILE_ASYNC \
-    --state_backend.rocks.incremental $STATE_BACKEND_ROCKS_INCREMENTAL \
-    --sequence_generator_source.sleep_time 15 \
-    --sequence_generator_source.sleep_after_elements 1"
-}
-
-BASE_JOB_CMD=`buildBaseJobCmd $ORIGINAL_DOP`
-
-JOB_CMD=""
-if [[ $SIMULATE_FAILURE == "true" ]]; then
-  # the submitted job should fail after at least 1 complete checkpoint.
-  # When simulating failures with the general purpose DataStream job,
-  # we disable restarting because we want to manually do that after the job fails.
-  JOB_CMD="$BASE_JOB_CMD \
-    --test.simulate_failure true \
-    --test.simulate_failure.num_records 200 \
-    --test.simulate_failure.num_checkpoints 1 \
-    --test.simulate_failure.max_failures 1 \
-    --environment.restart_strategy no_restart"
-else
-  JOB_CMD=$BASE_JOB_CMD
-fi
-
-DATASTREAM_JOB=$($JOB_CMD | grep "Job has been submitted with JobID" | sed 's/.* //g')
-
-if [[ $SIMULATE_FAILURE == "true" ]]; then
-  wait_job_terminal_state $DATASTREAM_JOB FAILED
-else
-  wait_job_running $DATASTREAM_JOB
-  wait_num_checkpoints $DATASTREAM_JOB 1
-  wait_oper_metric_num_in_records SemanticsCheckMapper.0 200
-
-  cancel_job $DATASTREAM_JOB
-fi
-
-# take the latest checkpoint
-CHECKPOINT_PATH=$(find_latest_completed_checkpoint ${CHECKPOINT_DIR}/${DATASTREAM_JOB})
-
-if [ -z $CHECKPOINT_PATH ]; then
-  echo "Expected an externalized checkpoint to be present, but none exists."
-  exit 1
-fi
-
-echo "Restoring job with externalized checkpoint at $CHECKPOINT_PATH ..."
-
-BASE_JOB_CMD=`buildBaseJobCmd $NEW_DOP "-s file://${CHECKPOINT_PATH}"`
-JOB_CMD=""
-if [[ $SIMULATE_FAILURE == "true" ]]; then
-  JOB_CMD="$BASE_JOB_CMD \
-    --test.simulate_failure true \
-    --test.simulate_failure.num_records 0 \
-    --test.simulate_failure.num_checkpoints 0 \
-    --test.simulate_failure.max_failures 0 \
-    --environment.restart_strategy no_restart"
-else
-  JOB_CMD=$BASE_JOB_CMD
-fi
-
-DATASTREAM_JOB=$($JOB_CMD | grep "Job has been submitted with JobID" | sed 's/.* //g')
-
-if [ -z $DATASTREAM_JOB ]; then
-    echo "Resuming from externalized checkpoint job could not be started."
-    exit 1
-fi
-
-wait_job_running $DATASTREAM_JOB
-wait_oper_metric_num_in_records SemanticsCheckMapper.0 200
-
-# if state is errorneous and the general purpose DataStream job produces alerting messages,
-# output would be non-empty and the test will not pass
+run_test_with_timeout 900 run_resume_externalized_checkpoints ${ORIGINAL_DOP} ${NEW_DOP} ${STATE_BACKEND_TYPE} ${STATE_BACKEND_FILE_ASYNC} ${STATE_BACKEND_ROCKS_INCREMENTAL} ${SIMULATE_FAILURE}
```

Comment:
`${STATE_BACKEND_FILE_ASYNC}` can be removed - the FsStateBackend is always async now.

Reply:
It looks as if `FsStateBackend` still has the corresponding code paths and there is also the `SyncSnapshotStrategySynchronicityBehavior` which is used by the `HeapKeyedStateBackend`. Given this, I would like to not do this clean up as part of this PR since it might require a bit more investigation.